text/3283-backward-compatible-default-features.md

```diff
@@ -0,0 +1,144 @@
+users to opt out of more features, because adding new default-enabled features
+for these would break existing users with `default-features = false`.
+
+To fix this, I propose we add the notion of a feature `base`. Users of crates
+have to select at least one of its base features. This means that in order to
+make another part of a crate optional, people can simply add a new optional
+feature and add it to all existing feature bases. Then they can create another
+feature base that excludes this new optional feature. The syntax of
+`default-features = false` will get replaced by `base = "<feature base>"`, with
+the default being `base = "default"`.
+
+In order to introduce this in a way that works with the existing
+`default-features = false` syntax, `default-features = false` will imply a
+feature base, `no-default-features`.
+
+For example, the warp crate enables HTTP2 by default and would like to make
+that optional. This was done in a PR, but the backward incompatibility was
+noticed (https://github.com/seanmonstar/warp/issues/913) and the PR was
+reverted. With this feature proposal, they could instead add a base called
+`no-default-features = ["http2"]`, so old crates depending on warp still get
+HTTP2 support. People can then opt out of the HTTP2 dependency by specifying a
+feature base like `base = "minimal_v1"`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When wanting to give users of your crate the possibility to opt out of parts of
+the functionality, you can turn to feature bases: Say you have a dependency on
+`large_crate`:
+
+```toml
+[dependencies]
+large_crate = "3.4"
+```
+
+You want to make that dependency optional, but since this would take away parts
+of the functionality of your crate as well, existing users must continue to use
+your crate with that dependency. The solution is to create a new feature base:
+
+```toml
+[package]
+feature-bases = ["minimal_v1"] # "default" is automatically a feature base
+
+[dependencies]
+large_crate = { version = "3.4", optional = true }
+
+[features]
+default = ["large_crate"]
+minimal_v1 = []
+```
+
+Users of your crate can now opt out of the large crate by specifying the
+following in their `Cargo.toml`:
+
+```toml
+[dependencies]
+your_crate = { version = "5.7", base = "minimal_v1" }
+```
+
+## Migration from default-features
+[migration-from-default-features]: #migration-from-default-features
+
+The `default-features = false` feature allowed to do the same migration, but
+only once per major version of a crate. To be backward-compatible with that
+approach, `default-features = false` implies the feature base called
+`default-features-false`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A new key `package.feature-bases` is introduced. This key specifies a list of
+features that, augmented with `default` and `default-features-false`, are
```

Comment:
This implies new semantics to a feature name that was not previously reserved.  What, if any, backwards compatibility concerns do we have about this?